File: src/main/java/easy/BinarySearch.java
package easy;

public class BinarySearch {

    /*
    Given an array of integers nums which is sorted in ascending order, and an integer target,
    write a function to search target in nums.
     If target exists, then return its index. Otherwise, return -1.
     */
    public static void main(String args[]) {
        System.out.println(search(new int[]{2, 5, 6}, 6));
    }

    public static int search(int[] nums, int target) {
        return searchb(nums, 0, nums.length - 1, target);
    }


    private static int searchb(int[] nums, int lo, int hi, int target) {

        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] == target) return mid;

            if (nums[mid] > target) {
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }

        }
        return lo;
    }


}

File: src/main/rust/BinarySearch.rs

/*
Given an array of integers nums which is sorted in ascending order
and an integer target, write a function to search target in nums. If target exists,
then return its index. Otherwise, return -1.
*/

impl Solution {
    pub fn search( nums: Vec<i32>, target: i32) -> i32 {

        return Self::sb(&nums, 0 , nums.len()-1, target);
    }


    fn sb (nums: &[i32], lo:usize, hi:usize, target: i32) -> i32 {


        let mut mid =   (lo+hi )/2;

        if (nums[mid]==target) {
            return  mid as i32;
        }

        if (lo<hi){
            if (nums[mid]<target) {

                return  Self::sb(nums, mid+1, hi , target);
            }else  {
                if (mid==0) {
                    return -1;
                }
                return  Self::sb(nums,0, mid-1 , target);
            }
        }
        return -1;
    }
}